src/generateResponseType.js

```diff
@@ -338,48 +357,59 @@ export const unionOrInterfaceToFlow = (
           }),
     selections: Selections,
 ): BabelNodeFlowType => {
-    const selectedAttributes: Array<
-        Array<BabelNodeObjectTypeProperty | BabelNodeObjectTypeSpreadProperty>,
-    > = type.possibleTypes
+    const allFields = selections.every(
+        (selection) => selection.kind === 'Field',
+    );
+    const selectedAttributes: Array<{
+        attributes: Array<
+            BabelNodeObjectTypeProperty | BabelNodeObjectTypeSpreadProperty,
+        >,
+        typeName: string,
+    }> = type.possibleTypes
         .slice()
         .sort((a, b) => {
             return a.name < b.name ? -1 : 1;
         })
         .map((possible) => {
             let seenTypeName = false;
-            return selections
-                .map((selection) =>
-                    unionOrInterfaceSelection(
-                        config,
-                        type,
-                        possible,
-                        selection,
-                    ),
-                )
-                .flat()
-                .filter((type) => {
-                    // The apollo-utilities "addTypeName" utility will add it
-                    // even if it's already specified :( so we have to filter out
-                    // the extra one here.
-                    if (
-                        type.type === 'ObjectTypeProperty' &&
-                        type.key.name === '__typename'
-                    ) {
-                        if (seenTypeName) {
-                            return false;
+            return {
+                attributes: selections
+                    .map((selection) =>
+                        unionOrInterfaceSelection(
+                            {
+                                ...config,
+                                path: allFields
+                                    ? config.path
+                                    : config.path.concat([possible.name]),
```

Comment:
Is this line hit when there's a fragment?

Reply:
Yeah, I don't really understand what this is doing. Maybe the problem is that I don't know what `kind` of `selection` there is other than `Field`.